### ISSUE TITLE
It's not easy to understand what is Reflective Design

### ISSUE THREAD

**Issue by AxCoder** [arlobelshee/AgileEngineeringFluency#6]
- The description is "Opposite to predictive design" while name is sound positive ("reflective" not "unpredictive")
- There is no definition on "predictive design".
- Hard to find the term in internet search 

Proposal:
- Write a positive definition of "reflective design"
- At least make a link to "Predictive design" with description

**Comment by JayBazuzi** [comment]
Great!

Do you have a description you like? Go ahead and send a pull request.

I'm guessing not, or you would have done it already. :-)

**Comment by aikiden** [comment]
Emergent Design?

On Thu, Apr 14, 2016 at 3:00 PM, Jay Bazuzi <notifications@github.com>
wrote:

> Great!
>
> Do you have a description you like? Go ahead and send a pull request.
>
> I'm guessing not, or you would have done it already. :-)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/arlobelshee/AgileEngineeringFluency/issues/6#issuecomment-210172917>
>



-- 
Dennis Britton
Innovation Accelerator
+1408.457.4978 Cell

**Comment by AxCoder** [comment]
I have an intuitive thought like. "Design that reflects domain area most straightforward way"  something like DDD. But not sure I users what Arlo means.

Отправлено с моего Windows Phone
________________________________

**Comment by aikiden** [comment]
What is the developer/team reflecting on?

On Thu, Apr 14, 2016 at 8:03 PM, Maxim Belugin <notifications@github.com>
wrote:

> I have an intuitive thought like. "Design that reflects domain area most
> straightforward way" something like DDD. But not sure I users what Arlo
> means.
>
> Отправлено с моего Windows Phone
> ________________________________
> From: Jay Bazuzi<mailto:notifications@github.com>
> Sent: ‎15.‎04.‎2016 1:00
> To: arlobelshee/AgileEngineeringFluency<mailto:
> AgileEngineeringFluency@noreply.github.com>
> Cc: Maxim Belugin<mailto:belugin@outlook.com>
> Subject: Re: [arlobelshee/AgileEngineeringFluency] It's not easy to
> understand what is Reflective Design (#6)
>
> Great!
>
> Do you have a description you like? Go ahead and send a pull request.
>
> I'm guessing not, or you would have done it already. :-)
>
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
>
> https://github.com/arlobelshee/AgileEngineeringFluency/issues/6#issuecomment-210172917
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/arlobelshee/AgileEngineeringFluency/issues/6#issuecomment-210263899>
>



-- 
Dennis Britton
Innovation Accelerator
+1408.457.4978 Cell

**Comment by arlobelshee** [comment]
General intent: reflective design is the technique of looking for insights in the code and asking it how it would like to have been designed, then updating the design to match. Where predictive design happens when code is written, reflective design happens when code is read.

Evolutionary design, on the other hand, is the combination of the two. Always design based on only the information currently known. Use predictive design techniques when writing, but mostly to aim for simplicity. Use reflective design techniques when reading, to extract intent.

DDD, I find, describes the final state of design, where Predictive / Reflective / Evolutionary describe the process of design transformation. You can use any of the three design processes to implement a domain-centric design. And you can use any of the three processes to implement an obtuse, highly-technical design.

**Comment by JayBazuzi** [comment]
It might be interesting to look at some examples. 

**Reflective Design**

I poked around on GitHub and found this `if/elseif/elseif` in a method called `ImportProject`: https://github.com/mono/MonoGame/blob/25578d0d8f4b5c0dcaec1e58f6df3c3016515b1e/Tools/Pipeline/Common/PipelineProjectParser.cs#L374

I've never read this code before, and don't really know what it does, but I my design sense suggests extracting three methods from the bodies of these blocks. The result would look something like this:

```
                        if (buildAction.Equals("Reference"))
                        {
                            ImportProject_BuildAction_Reference(...);
                        }
                        else if (buildAction.Equals("Content") || buildAction.Equals("None"))
                        {
                            ImportProject_BuildAction_ContentOrNone(...);
                        }
                        else if (buildAction.Equals("Compile"))
                        {
                            ImportProject_BuildAction_Compile(...);
                        }
```

This is an example of the [None->Honest naming transition](http://arlobelshee.com/tag/naming-is-a-process/). 

It's Reflective Design because someone wrote it one way, and then the need for those three methods became visible. 

I'm looking at the `ReadInclude*` methods and noticing that they are each used exactly once and have a lot of `out` parameters, which means they may not be the right methods. And they also have a lot of code in common. So I want to look for a way to extract the commonality of those methods in to a new place (eventually a new class), and the inline what is left. There are plenty more refactorings that want to be applied here, but that's fine for now: my goal is not Good, it's Better (fast). 

Someone doing **Predictive Design** when first writing this code might have thought "I bet there are multiple behaviors here and each has some complexity. I'll use Strategy Pattern.":

```
	ProjectImporter.GetStrategy(buildAction)
	               .Import(...);

	abstract class ProjectImporter
	{

		ProjectImporter GetStrategy(string buildAction)
		{
			var type = this.GetNestedTypes.SingleOrDefault(_ => _.Name == buildAction) ?? typeof(Nop);
			return (ProjectImporter) Activator.CreateInstance(type);
		}

		abstract void Import(...);

		class Reference : ProjectImporter
		{
		
		}
```

Is this better? I'm not sure, but it sure was fun to write!

**Evolutionary Design**

The original author could anticipate the complexity, and write the method calls I refactored to in the first iteration, even before implementing those methods. Later someone can notice the duplication between those methods and look for ways to eliminate it.
